test/extended/pods/graceful-shutdown.go

```diff
@@ -0,0 +1,457 @@
+		ctx := context.Background()
+
+		createTestBed(ctx, oc)
+
+		By("getting first worker node", func() {
+			nodes, err := oc.KubeClient().CoreV1().Nodes().List(ctx, metav1.ListOptions{LabelSelector: "node-role.kubernetes.io/worker="})
+			Expect(err).NotTo(HaveOccurred())
+			Expect(nodes.Items).NotTo(HaveLen(0))
+			node = &nodes.Items[0]
+		})
+
+		By("creating two pods with in range and out of range of grace periods", func() {
+			// Currently hard coded values
+			// TODO: Add dynamic support for current existing kubelet config.
+			// TODO: Maybe read rendered config ex: renderedConfigName := node.Labels["machineconfiguration.openshift.io/currentConfig"]
+			successPod = gracefulPodSpec(successPodName, node.Name, time.Minute*2)
+			errorPod = gracefulPodSpec(errorPodName, node.Name, time.Minute*10)
+			// Create two test pods
+			_, err := oc.KubeClient().CoreV1().Pods(namespace).Create(ctx, successPod, metav1.CreateOptions{})
+			if !apierrors.IsAlreadyExists(err) {
+				Expect(err).NotTo(HaveOccurred())
+			}
+
+			_, err = oc.KubeClient().CoreV1().Pods(namespace).Create(ctx, errorPod, metav1.CreateOptions{})
+			if !apierrors.IsAlreadyExists(err) {
+				Expect(err).NotTo(HaveOccurred())
+			}
+			// Wait for pods to be running
+			_, err = exutil.WaitForPods(
+				oc.KubeClient().CoreV1().Pods(namespace),
+				labels.SelectorFromSet(testPodLabel),
+				exutil.CheckPodIsRunning, 2, time.Second*60)
+			Expect(err).NotTo(HaveOccurred(), "unable to wait for pods")
+		})
+
+		By("triggering node reboot", func() {
+			// Reboot node
+			err := triggerReboot(oc, node.Name)
+			if !apierrors.IsAlreadyExists(err) {
+				Expect(err).NotTo(HaveOccurred())
+			}
+			// Wait for node to be rebooted before examining with debug pod
+			err = waitForNode(oc, node.Name)
+			Expect(err).NotTo(HaveOccurred(), "unable to watch node for status ready")
+		})
+
+		By("creating debug pod to gather files on the node", func() {
+			// We only need to query the node once and `ls` the desired directory.
+			// We can then determine the correct behavior by which file was written.
+			err := wait.Poll(time.Second, time.Second*60, func() (done bool, err error) {
+				outputString, err = debugPod(oc, node.Name)
+				// If output contains the below errors, retry
+				if err != nil && strings.Contains(outputString, "unable to create the debug pod") {
+					return false, nil
+				}
+				return true, err
+			})
+			Expect(err).NotTo(HaveOccurred(), "was not able to ls directory %s", hostPath)
+		})
+
+		Expect(outputString).To(ContainSubstring("%s", successPodName))
+
+		Expect(outputString).NotTo(ContainSubstring("%s", errorPodName))
+	})
+
+	AfterEach(func() {
+		deleteTestBed(context.Background(), oc)
+	})
+})
+
+func filePathForPod(podName string) string {
+	return fmt.Sprintf("%s/completed-%s", hostPath, podName)
+}
+
+// Wait for the node to come back up after reboot before attempting to validate results
+func waitForNode(oc *exutil.CLI, name string) error {
```

Comment:
Any particular reason you can't just use `framework.WaitForNodeToBeReady` from upstream's framework?

Reply:
Oh I didn't know that existed, didn't see it in the framework package. Created a new PR to use it instead, try and limit duplication https://github.com/openshift/origin/pull/27882 Thanks @soltysh !